### DISCUSSION TITLE
Flashing Artifacts Appearing on InvalidateVisual Call in Custom Control

### DISCUSSION THREAD

**Discussion by buchmiet** [AvaloniaUI/Avalonia#13291] [Q&A]
Hi!

I have developed a custom control in my Avalonia application, which functions as a multi-colored text block. The control has its own rendering logic defined. While the control performs as expected, I am encountering issues when trying to invalidate the control after appending a string to it. Below is the rendering method for my custom control:

```
  public override void Render(DrawingContext context)
  {
      charsperline = (Bounds.TopRight.X - Bounds.TopLeft.X) / charWidthActual;
      int counter = 0;
      int lines = 0;
      double startingChar;

      Avalonia.Point punkt = new Avalonia.Point();
      while (counter < Bufor.Count)
      {
          startingChar = 0;
          while (counter < Bufor.Count && startingChar < charsperline)
          {
              if (Bufor[counter].Value != '\r')
              {
                  punkt = new Avalonia.Point(startingChar * charWidthActual, (lines * charheight));
                  context.DrawText(Bufor[counter].Key, punkt);
                  startingChar++;
              }
              else
              {
                  if (counter < Bufor.Count && (Bufor[counter + 1].Value == '\n'))
                  {
                      counter++;
                      startingChar = 0;
                      lines++;
                  }
              }
              counter++;
          }
          lines++;
      }
  }
```
  
  The rendering logic works well and it renders the state of the buffer (**Bufor**) correctly. The **Bufor** is defined as follows (it seems you can not access the Text property of a formattedtext):
  
`  public List<KeyValuePair<FormattedText, char>> Bufor = new List<KeyValuePair<FormattedText, char>>();`

  I have a method AddString to append a new string to the Bufor:
  
```
    public void AddString(string inputString, int kolor)
       {
           for (int i = 0; i < inputString.Length; i++)
           {
               Bufor.Add
                   (
                       new KeyValuePair<FormattedText, char>
                           (
                               new FormattedText
                                   (
                                       inputString[i].ToString(),
                                       CultureInfo.InvariantCulture,
                                       FlowDirection.LeftToRight,
                                       ThemeTypeFace,
                                       charwidth,
                                       colors[kolor]
                                   ),
                               inputString[i]
                           )
                   );
           }
           Dispatcher.UIThread.InvokeAsync(InvalidateVisual, DispatcherPriority.Background);
       }
```
       
After calling InvalidateVisual through the Dispatcher.UIThread.InvokeAsync method, my application starts acting funny. Specifically, I observe flashing artifacts appearing over the entire application window, not just the control. These flashes last for a fraction of a second; sometimes revealing what's underneath the window, and sometimes displaying random colorful shapes as if drawingcontext was shown before it was cleared, yet Render method is not called.

It seems like there is a synchronization issue somewhere, but I'm unable to pinpoint the cause. I tried implementing my own ICustomDrawOperation, but the issue persists.

Interestingly, when I attempt to record the behavior using a screen recording application, the issue disappears, and the app window behaves perfectly fine.

The flashes do not correlate with the actual invalidation calls, as the control remains static most of the time until a new string is added. Even though nothing is being invalidated (post the initial invalidation call), and the Render method isn't being invoked, the flashing artifacts continue to appear once every now and then.

I'm wondering if my approach of invalidating the visual from another method is flawed, the entire approach is invalid, or if there's something else I'm missing. How should I go about diagnosing and resolving this issue? Any insights or suggestions would be greatly appreciated.

Thanks for ur help!

**Comment by timunie** [comment]
debug or release mode?

**Reply by buchmiet** [reply]
Both, Debug and Release alike. Is this the best way to call InvalidateVisual?

 Dispatcher.UIThread.InvokeAsync(InvalidateVisual(); , DispatcherPriority.Background); Again, I have tried recording those "flashes" but as soon as I have selected the area over the window to record and pressed REC, app had "stabilised".
 
 Could Invalidating the visual of the control somehow affect the entire application, which results in the occasional skip of a frame in rendering (hence the flashes of the background)?

**Reply by timunie** [reply]
make a minimal sample, attach it here for us to double-check on our machine.
